packages/frolint/src/utils/isInstanceOfNodeError.ts

```diff
@@ -0,0 +1,11 @@
+/* globals NodeJS */
+
+/**
+ * This is necessary when we want to use information available on a Nodejs specific `ErrnoException` object.
+ * Since the runtime instance is one from just `Error`, this function helps us narrowing at the type level.
+ */
+export function isInstanceOfNodeError<T extends new (...args: any) => Error>(
+  value: unknown
+): value is InstanceType<T> & NodeJS.ErrnoException {
+  return value instanceof Error && ("code" in value || "errno" in value);
```

Comment:
Hmm, as `ErrnoException` is an interface, we would need to do something like this for 100% accuracy:

```javascript
if (!(value instanceof Error)) return false;
try {
  const { errno, code, path, syscall, stack } = value;
  if (typeof errno !== "undefined" && typeof errno !== "number") return false;
  if (typeof code !== "undefined" && typeof code !== "string") return false;
  if (typeof path !== "undefined" && typeof path !== "string") return false;
  if (typeof syscall !== "undefined" && typeof syscall !== "string") return false;
  if (typeof stack !== "undefined" && typeof stack !== "string") return false;
} catch (_err) {
  // A getter failed
  return false;
}
```

Yes we don't need that accuracy, but why do we specifically check `code` and `errno` then? I think most instances of `Error` naturally follow the `NodeJS.ErrnoException` interface (mostly by the lack of the properties) so there's no practical need to check these properties.

Reply:
Well the only code paths we have right now use `code`. I thought `errno` could be somewhat useful as well.
I completely agree that the solution is partial, just mostly decided on it to solve my problem rather than have complete accuracy.

> I think most instances of Error naturally follow the NodeJS.ErrnoException interface (mostly by the lack of the properties) so there's no practical need to check these properties.

I'm not really sure what you're suggesting here. Just cast the `Error` we get to `ErrnoException`?

Reply:
> Just cast the `Error` we get to `ErrnoException`?

Either:

- Return `value is Error` from `isInstanceOfNodeError` and do casting at the call sites.
- Or, return `value is ErrnoException` and just remove the check. TypeScript just trusts the implementation when `is` or `asserts ... is` type is used.

Reply:
I see.
My issue, probably because I'm not familiar enough with Node.js, is that I thought there could be both `Error` and `ErrnoException` in a common program. The `Error` for the JS related errors (missing properties, TypeError, etc), and `ErrnoException` for Node.js API (readFile, etc). So having a simple check to differentiate the two at runtime would be important.
But actually all errors when running in Node.js are using `ErrnoException`?

Reply:
This is `ErrnoException`:

```typescript
interface ErrnoException extends Error {
    errno?: number;
    code?: string;
    path?: string;
    syscall?: string;
    stack?: string;
}
```

So, unless someone puts values of expected types to these properties, all `Error`s naturally implement `ErrnoException`.

Reply:
Ah, because all properties are `Partial`! I didn't realize, makes sense.
Just FYI, I looked into a REPL and things like `TypeError` don't have a `code` or `errno`.